### PR TITLE
feat: initial MVP application

### PR DESCRIPTION
**Pull request by erikgb**
This PR (when merged) should hopefully produce an application that we can try out! There are probably a lot of things to improve, but let's get started! 🤠

### CODE REVIEW

File: src/main/java/io/statnett/k3a/topicterminator/strategy/InternalTopic.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+        return allNames.stream()
+            .filter(this::isInternal)
+            .collect(Collectors.toSet());
+    }
+
+    private boolean isInternal(String name) {
+        return isUsedByConnect(name) || isUsedInternal(name) || isUsedByStream(name);
+    }
+
+    private boolean isUsedInternal(String name) {
```
Comment (sverrehu) [inline src/main/java/io/statnett/k3a/topicterminator/strategy/InternalTopic.java:32]: Should this be `isUsedInternally`?

**Review by sverrehu — DISMISSED** [review]

**Comment by sverrehu** [comment]
I think we should have a walkthrough on this on Monday.

File: src/main/java/io/statnett/k3a/topicterminator/strategy/InternalTopic.java (hunk 1 above)
Comment (erikgb) [reply src/main/java/io/statnett/k3a/topicterminator/strategy/InternalTopic.java:30]: I think the definition of an "internal" topic must be defined better. I am struggling to find reasonable docs in this area.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+RUN mvn --batch-mode -Dmaven.test.skip=true package
+
+FROM eclipse-temurin:21.0.4_7-jre-alpine
+WORKDIR /app
+COPY --from=builder /workspace/target/k3a-topic-terminator.jar ./
+RUN apk update \
+  && apk upgrade \
+  && rm -rf /var/cache/apk/*
+
+ENTRYPOINT ["java","-jar","k3a-topic-terminator.jar"]
```
Comment (sverrehu) [inline Dockerfile:16]:
```suggestion
ENTRYPOINT ["java", "-jar", "k3a-topic-terminator.jar"]
```

File: src/main/java/io/statnett/k3a/topicterminator/ApplicationProperties.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+     *
+     * @see <a href="https://en.wikipedia.org/wiki/ISO_8601#Durations">ISO-8601 duration format</a>
+     */
+    @NotEmpty
+    private String fixedRateString;
+
+    /**
+     * If set to `true` the application won't delete anything,
+     * but just log the topics that would have been deleted if
+     * dry-run disabled.
```
Comment (sverrehu) [inline src/main/java/io/statnett/k3a/topicterminator/ApplicationProperties.java:23]:
```suggestion
     * dry-run was disabled.
```

File: src/main/java/io/statnett/k3a/topicterminator/TopicTerminator.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+            );
+
+            for (ReservedTopic reservedTopic : reservedTopics) {
+                unusedTopics.removeAll(reservedTopic.getNames(client));
+            }
+
+            if (props.isDryRun()) {
+                unusedTopics.forEach(t -> log.info("Topic {} is considered unused and would be deleted in non dry-run mode", t));
+            } else {
+                log.info("{} topic(s) to be deleted: ", unusedTopics.size());
```
Comment (sverrehu) [inline src/main/java/io/statnett/k3a/topicterminator/TopicTerminator.java:53]:
```suggestion
            log.info("{} topic(s) to be deleted: ", unusedTopics.size());
            if (props.isDryRun()) {
                unusedTopics.forEach(t -> log.info("Topic {} is considered unused and would be deleted in non dry-run mode", t));
            } else {
```

File: src/main/java/io/statnett/k3a/topicterminator/strategy/ConsumedTopic.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+
+/**
+ * Topic that is consumed.
+ * <p>
+ * If a consumer group have not consumed anything for a while,
```
Comment (sverrehu) [inline src/main/java/io/statnett/k3a/topicterminator/strategy/ConsumedTopic.java:16]:
```suggestion
 * If a consumer group has not consumed anything for a while,
```

File: src/main/java/io/statnett/k3a/topicterminator/strategy/InternalTopic.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+        return name.startsWith("_");
+    }
+
+    private boolean isUsedByConnect(String name) {
+        return name.startsWith("stream.connect");
+    }
+
+    private boolean isUsedByStream(String name) {
+        return name.endsWith("-changelog") || name.endsWith("-repartition") || name.endsWith("-rekey");
```
Comment (sverrehu) [inline src/main/java/io/statnett/k3a/topicterminator/strategy/InternalTopic.java:39]: Would be nice if we could clean out these based on the existence of the main topic. If at all doable. :-)

File: src/test/java/io/statnett/k3a/topicterminator/ApplicationTest.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+    void testTerminateUnusedTopics() throws Exception {
+        // Put some data on topic-with-data topic
+        kafkaTemplate.send(TOPIC_WITH_DATA, "foo").get();
+
+        // Wait until consumer is started and registered in cluster
+        try (AdminClient client1 = AdminClient.create(kafkaAdmin.getConfigurationProperties())) {
+            await().until(() -> client1.listConsumerGroupOffsets(GROUP_CONSUMER)
+                .partitionsToOffsetAndMetadata().get().keySet().stream()
+                .anyMatch(topicPartition -> topicPartition.topic().equals(TOPIC_CONSUMED)));
+
```
Comment (sverrehu) [inline src/test/java/io/statnett/k3a/topicterminator/ApplicationTest.java:53]:
```suggestion
```

File: src/main/java/io/statnett/k3a/topicterminator/strategy/InternalTopic.java (hunk 6 above)
Comment (erikgb) [reply src/main/java/io/statnett/k3a/topicterminator/strategy/InternalTopic.java:39]: I agree! I think the "internal topics" is something we should look at pretty soon.

**Review by sverrehu — APPROVED** [review]

**Pull request merged**